### PR TITLE
add rpc server to the oculus module to control finite state machine of the module

### PR DESCRIPTION
**Pull request by kouroshD**
Problem: 
Especially by using the VIVE, it is often tricky to distinguish the left from the right hand. It would be better and safer to start the teleoperation pipeline from a terminal command
Solution:
An RPC command is added to the oculus module.

### CODE REVIEW

**Comment by kouroshD** [comment]
@S-Dafarra  @GiulioRomualdi  if you can test this feature for the demo, it would be great. I could build it without any problem, but I could not test it.

**Comment by S-Dafarra** [comment]
I quickly checked it. Here a couple of comments:
1. the ``RPC`` answer is asynchronous, hence you need to use some ``mutex`` around.
2. I would suggest using separate methods dealing with each command.
3. Instead of using the ``respond`` method, you can specify a thrift with all the commands you need (http://www.yarp.it/git-master/thrift_tutorial_simple.html). This has the following advantages:
    - for each command it automatically generates a virtual method that you can implement, thus automatically addressing point 2 above;
    - it handles automatically the ``help`` command;
    - it handles automatically more complex inputs, like vectors
    - if you write some documentation, this is automatically reported when typing ``help myCommand``
    - it is more scalable; it is easier to add/change/remove commands

You can find some example of usage of the thrifts in the ``walking-controller``:
- the thrift file is https://github.com/robotology/walking-controllers/blob/75429cbbf444c92496beb863667cac0c50eb542e/src/WalkingModule/thrifts/WalkingCommands.thrift
- this is the command to launch in the CMakeLists https://github.com/robotology/walking-controllers/blob/75429cbbf444c92496beb863667cac0c50eb542e/src/WalkingModule/CMakeLists.txt#L37-L38 this will automatically create some files that you need to add to the executable https://github.com/robotology/walking-controllers/blob/75429cbbf444c92496beb863667cac0c50eb542e/src/WalkingModule/CMakeLists.txt#L42.
- In the module then, you include the autogenerated header https://github.com/robotology/walking-controllers/blob/75429cbbf444c92496beb863667cac0c50eb542e/src/WalkingModule/include/WalkingControllers/WalkingModule/Module.h#L53. This has the same name of the thrift file. 
- you inherit from the autogenerated class (that has the same name of the thrift file): https://github.com/robotology/walking-controllers/blob/75429cbbf444c92496beb863667cac0c50eb542e/src/WalkingModule/include/WalkingControllers/WalkingModule/Module.h#L61
- You implement the methods related to the commands: https://github.com/robotology/walking-controllers/blob/75429cbbf444c92496beb863667cac0c50eb542e/src/WalkingModule/include/WalkingControllers/WalkingModule/Module.h#L258-L288
- You still have the RPC port https://github.com/robotology/walking-controllers/blob/75429cbbf444c92496beb863667cac0c50eb542e/src/WalkingModule/include/WalkingControllers/WalkingModule/Module.h#L120
- You use this commands to "attach" the methods you implemented to the RPC port https://github.com/robotology/walking-controllers/blob/75429cbbf444c92496beb863667cac0c50eb542e/src/WalkingModule/src/Module.cpp#L170-L176
- You close the port normally: https://github.com/robotology/walking-controllers/blob/75429cbbf444c92496beb863667cac0c50eb542e/src/WalkingModule/src/Module.cpp#L356

**Comment by kouroshD** [comment]
@S-Dafarra  I have updated the PR with the requested changes. Please have a look on it.
Thanks.

File: modules/Oculus_module/include/OculusModule.hpp [hunk 1]
```diff
@@ -244,6 +247,52 @@ class OculusModule : public yarp::os::RFModule
...
+     * 1. ("help");
+     * 2. ("prepare").
+     * 3. ("run")
+     * @param reply is the response of the server.
+     * 1. OK in case of success;
+     * 2. Not OK in case of failure.
+     * 3. list of available commands in case of "help" command
+     * @return true in case of success and false otherwise.
+     */
+    bool respond(const yarp::os::Bottle& command, yarp::os::Bottle& reply) final;
```
Comment (S-Dafarra) [inline modules/Oculus_module/include/OculusModule.hpp:288]: I think this is not necessary anymore

File: modules/Oculus_module/src/OculusModule.cpp [hunk 2]
```diff
@@ -553,6 +566,266 @@ bool OculusModule::close()
...
+        //                            + ", while it should be in state: "
+        //                            + getStringFromOculusState(OculusFSM::InPreparation));
+        //        }
+    } else
+    {
+        reply.addString("Entered command is incorrect, Enter `help` to know available commands");
+    }
+
+    return true;
+}
```
Comment (S-Dafarra) [inline modules/Oculus_module/src/OculusModule.cpp:713]: I think this should not be necessary anymore

File: modules/Oculus_module/src/OculusModule.cpp [hunk 3]
```diff
@@ -553,6 +566,266 @@ bool OculusModule::close()
...
+        yWarning() << "The teleoperation is in state: " + getStringFromOculusState(m_state)
+                          + ", while it should be in state: "
+                          + getStringFromOculusState(OculusFSM::InPreparation)
+                   << "in order to run the teleoperation.";
+        yWarning() << "No update on the teleoperation state, try again.";
+        return false;
+    }
+
+    return true;
+}
```
Comment (S-Dafarra) [inline modules/Oculus_module/src/OculusModule.cpp:746]: This is not doing anything

File: modules/Oculus_module/src/OculusModule.cpp [hunk 4]
```diff
@@ -967,19 +1240,12 @@ bool OculusModule::updateModule()
...
         {
-            // TODO add a visual feedback for the user
-            if (m_moveRobot)
-            {
-                cmd.addString("prepareRobot");
-                m_rpcWalkingClient.write(cmd, outcome);
-            }
-            m_state = OculusFSM::InPreparation;
-            yInfo() << "[OculusModule::updateModule] prepare the robot";
+            std::lock_guard<std::mutex> guard(m_mutex);
```
Comment (S-Dafarra) [inline modules/Oculus_module/src/OculusModule.cpp:1245]: The lock should not just be here. For example, in the lines above you are accessing to the ``m_state`` variable, that can be edited at anytime by the RPC calls. The scope of the lock should be on the entire ``updateModule``.

**Review by S-Dafarra — CHANGES_REQUESTED** [review]
I think the lock of the mutex is missing in other places, like the ``close`` method.

File: modules/Oculus_module/include/OculusModule.hpp (hunk 1 above)
Comment (kouroshD) [reply modules/Oculus_module/include/OculusModule.hpp:288]: Yes, I should remove it, it is a left over, thanks.

**Comment by kouroshD** [comment]
@S-Dafarra  Thank you for the comments, I have updated the PR with your comments.

File: modules/Oculus_module/src/OculusModule.cpp [hunk 5]
```diff
@@ -518,7 +531,10 @@ bool OculusModule::configure(yarp::os::ResourceFinder& rf)
         yInfo() << "[OculusModule::configure] Cameras have been reset.";
     }
 
-    m_state = OculusFSM::Configured;
+    {
+        std::lock_guard<std::mutex> guard(m_mutex);
```
Comment (S-Dafarra) [inline modules/Oculus_module/src/OculusModule.cpp:535]: I would lock the entire method. ``m_state`` is not the only variable that could be affected by concurrent read/write

File: modules/Oculus_module/src/OculusModule.cpp [hunk 6]
```diff
@@ -553,6 +569,149 @@ bool OculusModule::close()
...
+        yWarning() << "No update on the teleoperation state, try again.";
+        return false;
+    }
+
+    return this->runningModule();
+}
+
+bool OculusModule::preparingModule()
+{
+    std::lock_guard<std::mutex> guard(m_mutex);
```
Comment (S-Dafarra) [inline modules/Oculus_module/src/OculusModule.cpp:636]: This is a deadlock. If ``preparingModule`` is called by other methods, then it should not attempt to lock the mutex if the caller is also locking it

File: modules/Oculus_module/src/OculusModule.cpp [hunk 7]
```diff
@@ -553,6 +569,149 @@ bool OculusModule::close()
...
+    }
+    m_state = OculusFSM::InPreparation;
+    yInfo() << "[OculusModule::preparingModule] prepare the robot";
+
+    return true;
+}
+
+bool OculusModule::runningModule()
+{
+    std::lock_guard<std::mutex> guard(m_mutex);
```
Comment (S-Dafarra) [inline modules/Oculus_module/src/OculusModule.cpp:651]: This is also a deadlock

**Review by S-Dafarra — CHANGES_REQUESTED** [review]

**Comment by kouroshD** [comment]
@S-Dafarra  I have updated the PR with your suggestions. Moreover, I have rebased the code. Let me know if you have other comments as well. Thanks.

**Comment by kouroshD** [comment]
We have tested PR added to this branch :  https://github.com/robotology/walking-teleoperation/tree/modify_dcmWalkingRetargeting_rpc_Oculus
It was working as expected.
@S-Dafarra  If you can approve the PR, I will proceed with the merge.

**Review by S-Dafarra — APPROVED** [review]

**Comment by S-Dafarra** [comment]
Just one comment for future reference. When using the gloves now, we could avoid connecting to the ``JoypadControlServer``.

**Pull request merged**